### DISCUSSION TITLE
tests on macos run infinitely.

### DISCUSSION THREAD

**Discussion by root-11** [community/community#38364] [Actions] [answered]
Hi - I've just seen macos run forever:
![image](https://user-images.githubusercontent.com/4048422/200131371-e80ae3e8-2663-4429-a4cb-8b62a76684f7.png)

I have no idea why but consider it a waste of compute as the runner should finish with roughly the same amount of time as winos and linux.

My workflow file is here: https://github.com/root-11/tablite/actions/runs/3400502081/workflow

Help?

**Comment by root-11** [comment]
The run just finished:
![image](https://user-images.githubusercontent.com/4048422/200136628-3f4e8234-d930-4a3e-9374-8afe4ba3b1f6.png)

The code is identical for all builds, yet macos took 2 hours, in comparison to the others that took 15 minutes.

**Reply by david-engelmann** [reply]
What command takes so long, can you share the details of the build?

**Comment by root-11** [comment]
It's all python3.7+ and there are no unbounded loops in the code.

The build matrix is here:
https://github.com/root-11/tablite/blob/master/.github/workflows/python-package.yml 

The run is here:
https://github.com/root-11/tablite/actions/runs/3400882264

**Reply by david-engelmann** [reply]
The issue is occurring with one of the pytest. Next time it hangs, I would investigate the "Test with pytest" layer and see which test is taking so long. If it's an os related test, then it will be easier to investigate from there

**Comment by root-11** [comment]
Feel free to restart the job and/or clone the repo.
I'm a mathematician and mainly use python as a calculator on winos and
linux.
I've never used macos and wouldn't know where to start.

**Reply by david-engelmann** [reply]
I'll fork it, and make a commit now

**Reply by david-engelmann** [reply]
[I've added the time the pytest take as output and limited it to just py 3.9](https://github.com/david-engelmann/tablite/actions/runs/3405081916)

**Reply by david-engelmann** [reply]
the tests/test_filereader_performance.py::test01 was the longest by far for both linux and windows, still waiting on the mac runner to finish

**Reply by root-11** [reply]
That sounds correct.

**Reply by david-engelmann** [reply]
For sure, loads a TB tables. We will see if the mac runner is the same test and is just drastically slower or if another test gets hung.

**Reply by root-11** [reply]
Just looking at the logs from your test, winos and linux have roughly same order and time consumption. Macos looks completely different. Wow.

**Reply by david-engelmann** [reply]
#### Mac Build ####
| ExecutionTime | TestName                                                                      |
|---------------|-------------------------------------------------------------------------------|
| 719.18s call  | tests/test_filereader_performance.py::test01                                  |
| 65.95s call   | tests/test_imputation.py::test_replace_missing_values_04                      |
| 65.72s call   | tests/test_filter.py::test_filter                                             |
| 50.78s call   | tests/test_groupby_and_pivot.py::test_reverse_pivot                           |
| 31.63s call   | tests/test_groupby_and_pivot.py::test_groupby_w_pivot                         |
| 28.74s call   | tests/test_filereader_formats.py::test_long_texts                             |
| 17.91s call   | tests/test_api_basics.py::test03_verify_slice_operator_for_multitype_datasets |
| 17.07s call   | tests/test_filereader_formats.py::test_filereader_saptxt                      |
| 14.49s call   | tests/test_new_tutorial.py::test_sort_parallel                                |
| 13.70s call   | tests/test_memory_footprint.py::test_recreate_readme_comparison               |
| 13.05s call   | tests/test_filereader_formats.py::test_filereader_gdocs1xlsx_no_header        |
| 12.50s call   | tests/test_imputation.py::test_replace_missing_values_03                      |
| 11.48s call   | tests/test_new_tutorial.py::test_the_basics                                   |
| 11.05s call   | tests/test_filereader_formats.py::test_filereader_book1_txt_chunks_and_offset |
| 10.77s call   | tests/test_filereader_formats.py::test_filereader_book1_txt_chunks            |
| 10.72s call   | tests/test_new_tutorial.py::test_lookup_logic                                 |
| 9.78s call    | tests/test_imputation.py::test_replace_missing_values_06                      |
| 9.59s call    | tests/test_new_tutorial.py::test_group_by_logic                               |
| 9.08s call    | tests/test_task_manager.py::test_main                                         |
| 9.07s call    | tests/test_filereader_formats.py::test_filereader_csv_f12                     |

**Reply by david-engelmann** [reply]
I'd say it's not 1 particular test that is causing the longer run but a combo of a bunch of different test that take longer to execute.

**Reply by root-11** [reply]
tablite uses hdf5 and is quite heavy on the IOPS. Could this be the cause?

**Reply by david-engelmann** [reply]
Windows Build
| ExecutionTime | TestName                                                        |
|---------------|-----------------------------------------------------------------|
| 540.19s call  | tests/test_filereader_performance.py::test01                    |
| 22.42s call   | tests/test_filter.py::test_filter                               |
| 15.48s call   | tests/test_imputation.py::test_replace_missing_values_04        |
| 14.38s call   | tests/test_groupby_and_pivot.py::test_reverse_pivot             |
| 11.86s call   | tests/test_task_manager.py::test_main                           |
| 8.83s call    | tests/test_memory_footprint.py::test_recreate_readme_comparison |
| 8.45s call    | tests/test_new_tutorial.py::test_sort_parallel                  |

I would say that a theme in additional tests that are present in the ios build is the need for memory

**Reply by david-engelmann** [reply]
@root-11 Don't forget to mark an answer to close it out

**Reply by david-engelmann** [reply]
> @root-11 Don't forget to mark an answer to close it out

usually an answer, not the question :)

**Reply by root-11** [reply]
Sorry. It's just that you're replies don't have the option to "mark as answer". I think it's because we are inside a conversation?

**Comment by david-engelmann** [comment, marked as answer]
[Action job used to debug](https://github.com/david-engelmann/tablite/actions/runs/3405081916) :)
